server/api/submitflag.js

```diff
@@ -44,20 +45,29 @@ module.exports = {
 
     const challenge = challenges.getChallenge(challengeid)
 
-    if (challenge) {
-      if (submittedFlag === challenge.flag) {
-        const solved = await db.solves.getSolvesByUserIdAndChallId({ userid: uuid, challengeid: challengeid })
-        if (solved === undefined) {
-          await db.solves.newSolve({ id: uuidv4(), challengeid: challengeid, userid: uuid, createdat: new Date() })
-          return responses.goodFlag
-        } else {
-          return responses.badAlreadySolvedChallenge
-        }
-      } else {
-        return responses.badFlag
-      }
-    } else {
+    if (!challenge) {
       return responses.badChallenge
     }
+
+    const bufSubmittedFlag = Buffer.from(submittedFlag)
+    const bufCorrectFlag = Buffer.from(challenge.flag)
+
+    if (bufSubmittedFlag.length !== bufCorrectFlag.length) {
+      return responses.badFlag
+    }
+
+    if (!crypto.timingSafeEqual(bufSubmittedFlag, bufCorrectFlag)) {
+      return responses.badFlag
+    }
+
+    try {
+      await db.solves.newSolve({ id: uuidv4(), challengeid: challengeid, userid: uuid, createdat: new Date() })
+      return responses.goodFlag
+    } catch (e) {
+      if (e.constraint === 'uq') {
```

Comment:
might want to add a comment, but the rest of the code makes it fairly self explanatory